ws-server/src/server.rs

```diff
@@ -71,26 +82,48 @@ impl Server {
 		self.listener.local_addr().map_err(Into::into)
 	}
 
+	/// Returns the handle to stop the running server.
+	pub fn stop_handle(&self) -> StopHandle {
+		StopHandle { stop_sender: self.stop_pair.0.clone(), stop_handle: self.stop_handle.clone() }
+	}
+
 	/// Start responding to connections requests. This will block current thread until the server is stopped.
 	pub async fn start(self) {
-		let mut incoming = TcpListenerStream::new(self.listener);
+		// Lock the stop mutex so existing stop handles can wait for server to stop.
+		// It will be unlocked once this function returns.
+		let _stop_handle = self.stop_handle.lock().await;
+
+		let mut incoming = TcpListenerStream::new(self.listener).fuse();
 		let methods = Arc::new(self.methods);
 		let cfg = self.cfg;
 		let mut id = 0;
+		let mut stop_receiver = self.stop_pair.1;
 
-		while let Some(socket) = incoming.next().await {
-			if let Ok(socket) = socket {
-				socket.set_nodelay(true).unwrap_or_else(|e| panic!("Could not set NODELAY on socket: {:?}", e));
+		loop {
+			futures_util::select! {
+				socket = incoming.next() => {
+					if let Some(Ok(socket)) = socket {
+							socket.set_nodelay(true).unwrap_or_else(|e| panic!("Could not set NODELAY on socket: {:?}", e));
 
-				if Arc::strong_count(&methods) > self.cfg.max_connections as usize {
-					log::warn!("Too many connections. Try again in a while");
-					continue;
-				}
-				let methods = methods.clone();
+							if Arc::strong_count(&methods) > self.cfg.max_connections as usize {
+								log::warn!("Too many connections. Try again in a while");
+								continue;
+							}
+							let methods = methods.clone();
 
-				tokio::spawn(background_task(socket, id, methods, cfg));
+							tokio::spawn(background_task(socket, id, methods, cfg));
 
-				id += 1;
+							id += 1;
+					} else {
+						break;
```

Comment:
hmm, we shouldn't kill the entire server because the `incoming.next()` failed? 

Can you explain this?